Updated transaction state changes mechanism

gemdev111 opened the pull request:
Added adaptive polling & tests
Added logic to handle minChainBlockTime in polling
Added TransactionStateUpdateJob (sync balances, stake data and NFT state after a transaction is finalized)
Introduced TransactionStateService (isolates RPC state-fetch + DB-update logic)
TransactionStateUpdateJob now owns stateService and postProcessingService
TransactionService constructs both services and passes them to jobs
JobRunner extended with clock for accuracy
JobConfiguration extended with timeLimit
Updated retry-calculator tests

closes #478

gemdev111 committed:
Updated transaction state changes mechanism
Added adaptive polling & tests
Added logic to handle minChainBlockTime in polling

gemdev111 committed:
Minor refactoring

gemdev111 committed:
Update core

gemdev111 committed:
removed unused separator

gemdev111 committed:
Merge branch 'main'

gemdev111 committed:
transactionsPoller updated in more "generic" way to TimerPoller
Cleanup, naming updates
Enabled unit_frameworks tests for TransactionService

gemdev111 committed:
Merge branch 'main'

gemcoder21 commented on Services/TransactionService/Sources/TimerPoller/TimerPoller.swift:
```diff
@@ -0,0 +1,52 @@
+    }
+}
+
+extension Duration {
```
can you add to primitives?

gemcoder21 commented on Services/TransactionService/Sources/TimerPoller/TimerPollerConfiguration.swift:
```diff
@@ -0,0 +1,27 @@
+}
+
+extension TimerPollerConfiguration {
+    static let `default` = TimerPollerConfiguration(
```
let's not have default, let the caller specify

gemcoder21 commented on Services/TransactionService/Sources/TimerPoller/PollIntervalCalculator.swift:
```diff
@@ -0,0 +1,23 @@
+    static func nextInterval(
+        configuration: TimerPollerConfiguration,
+        currentInterval: Duration,
+        recommendedInterval: Duration
```
```suggestion
        maxInterval: Duration
```

gemdev111 replied:
TimerPollerConfiguration already has maxInterval, let's use `targetInterval` or left `recommendedInterval` for result ? 
```
struct TimerPollerConfiguration: Sendable {
    let maxInterval: Duration
    let idleInterval: Duration
    let stepFactor: Double

    init(
        maxInterval: Duration,
        idleInterval: Duration,
        stepFactor: Double
    ) {
        self.maxInterval = maxInterval
        self.idleInterval = idleInterval
        self.stepFactor = stepFactor
    }
}
```

```
struct PollIntervalCalculator {
    static func nextInterval(
        configuration: TimerPollerConfiguration,
        currentInterval: Duration,
        recommendedInterval: Duration
    ) -> Duration {
        let stepFactor = configuration.stepFactor
        let maxInterval = configuration.maxInterval
        let idle = configuration.idleInterval
        
        if recommendedInterval < idle {
            return recommendedInterval
        } else if recommendedInterval >= maxInterval {
            return maxInterval
        }
        
        return .min(currentInterval * stepFactor, maxInterval)
    }
}
```

gemcoder21 commented on Services/TransactionService/Sources/TimerPoller/PollResult.swift:
```diff
@@ -0,0 +1,16 @@
+
+    init(
+        isActive: Bool,
+        recommendedInterval: Duration
```
```suggestion
        maxInterval: Duration
```

gemcoder21 commented on Services/TransactionService/Sources/TransactionService.swift:
```diff
@@ -35,22 +35,37 @@ public struct TransactionService: Sendable {
 
     public func addTransactions(walletId: String, transactions: [Transaction]) throws {
         try transactionStore.addTransactions(walletId: walletId, transactions: transactions)
+        setup()
```
```suggestion
        setup()
```

why do we need to setup on adding new transaction?

gemcoder21 commented on Services/TransactionService/Sources/TransactionService.swift:
```diff
@@ -35,22 +35,37 @@ public struct TransactionService: Sendable {
+        // Check if any remain pending
+        let remainPendingTransactions = try transactionStore.getTransactions(state: .pending)
+        return PollResult(
+            isActive: !remainPendingTransactions.isEmpty,
```
```suggestion
            isActive: transactions.isNotEmpty,
```

gemcoder21 replied:
or `pendingTransactions.isNotEmpty`

gemdev111 committed:
Merge branch 'main'

gemdev111 committed:
Cleanup
Removed default static config TimerPollerConfiguration
Moved duraion extension to primitives
Minor naming updates

gemdev111 committed:
Merge branch 'main'

gemcoder21 committed:
Add Sendable conformance to view model structs

gemdev111 committed:
Separated TransactionService
Added TransactionPoller, responsible for poll cached pending trancastions  updates
Added TransactionUpdater
Minor updates on TimerPoller, now it canceled if no pending transactions

gemdev111 committed:
Update core

gemdev111 committed:
Merge branch 'main'

gemdev111 committed:
Updates to TransactionService
Added separate package JobRunner as an abstract poller with set of configurations
Added TransactionUpdateJob & TransactionUpdateService
Refactored TrasnactionService to support new polling logic

gemdev111 committed:
Merge branch 'main'

gemdev111 committed:
cleanup

gemdev111 committed:
downgrade JobRunner swift tools to 6.0

gemdev111 committed:
Update core

gemdev111 committed:
Merge branch 'main'

gemdev111 committed:
Merge branch 'main' into transactions-status-polling

gemcoder21 commented on Packages/JobRunner/Sources/JobRunner.swift:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    public func stopAll() {
+        tasks.forEach { $0.value.cancel() }
```
you need to use `cancelJob()` for each to avoid code duplication

gemcoder21 commented on Packages/JobRunner/Sources/Types/JobStatus.swift:
```diff
@@ -0,0 +1,9 @@
+
+import Foundation
+
+public enum JobStatus: Sendable {
```
for simplicity let's use success and failure only

gemcoder21 commented on Services/TransactionService/Sources/TransacitonUpdateJob.swift:
```diff
@@ -0,0 +1,29 @@
+import Primitives
+import JobRunner
+
+public struct TransactionUpdateJob: Job {
```
```suggestion
public struct TransactionStateUpdateJob: Job {
```

gemcoder21 commented on Services/TransactionService/Sources/TransactionService.swift:
```diff
@@ -20,31 +22,51 @@ public struct TransactionService: Sendable {
+            stakeService: stakeService,
+            nftService: nftService
+        )
+        self.runner = JobRunner(
```
can you move it up? to avoid init in init() function

gemdev111 committed:
Minor adjustments

gemdev111 committed:
Merge branch 'main'

gemdev111 committed:
Update core

gemdev111 committed:
Merge branch 'main'

gemdev111 committed:
Update core

gemdev111 committed:
Merge branch 'main'

gemdev111 committed:
Code cleanup
Added TransactionStateUpdateJob (sync balances, stake data and NFT state after a transaction is finalized)
Introduced TransactionStateService (isolates RPC state-fetch + DB-update logic)
TransactionStateUpdateJob now owns stateService and postProcessingService
TransactionService constructs both services and passes them to jobs
JobRunner extended with clock for accuracy
JobConfiguration extended with timeLimit
Updated retry-calculator tests

gemdev111 committed:
Merge branch 'main'

gemdev111 committed:
github CI fix

gemdev111 committed:
Fixed unit_frameworks tests

gemdev111 committed:
Removed unused comma, one more time github ci..

gemcoder21 committed:
Merge branch 'main' into transactions-status-polling

gemcoder21 committed:
Rename TransactionPostProcessingService to TransactionStateUpdatePostJob for consistency across services
